File: public/services/actions/ai.ts
import { http, Result } from "@fider/services/http"
import { AIMessage } from "@fider/models"

export interface AIFinalizeResponse {
  title: string
  description: string
  brief: string
  tags: string[]
}

export const aiIdeate = async (productId: number, messages: AIMessage[]): Promise<Result<{ reply: string; ready: boolean }>> => {
  return http.post<{ reply: string; ready: boolean }>(`/api/v1/ai/ideate`, { productId, messages })
}

export const aiFinalize = async (productId: number, messages: AIMessage[]): Promise<Result<AIFinalizeResponse>> => {
  return http.post<AIFinalizeResponse>(`/api/v1/ai/finalize`, { productId, messages })
}

export const getBriefTranscript = async (postNumber: number): Promise<Result<{ messages: AIMessage[] }>> => {
  return http.get<{ messages: AIMessage[] }>(`/_api/posts/${postNumber}/brief/transcript`)
}

export const getIdeaBrief = async (postNumber: number): Promise<Result<{ content: string; createdAt: string }>> => {
  return http.get<{ content: string; createdAt: string }>(`/api/v1/posts/${postNumber}/brief`)
}

export const updateAISettings = async (settings: {
  enabled: boolean
  provider: string
  apiKey: string
  model: string
  customBaseUrl: string
  customModel: string
  webSearchEnabled: boolean
  webSearchProvider: string
  webSearchApiKey: string
  webSearchBaseUrl: string
}): Promise<Result> => {
  return http.post(`/_api/admin/settings/ai`, settings)
}

export const getAIProviderKey = async (): Promise<Result<{ apiKey: string }>> => {
  return http.get<{ apiKey: string }>(`/_api/admin/settings/ai/key`)
}

export const upsertAIAgent = async (agent: {
  productId: number | null
  description: string
  instructions: string
  enabled: boolean
  webSearchEnabled: boolean
}): Promise<Result> => {
  return http.post(`/_api/admin/ai/agents`, agent)
}

File: app/models/cmd/ai.go
package cmd

import "github.com/getfider/fider/app/models/entity"

// SetAISettings stores the tenant's provider configuration. An empty APIKey
// means "keep the stored key" — the key is write-only from the client's view.
type SetAISettings struct {
	Enabled       bool
	Provider      string // "claude" | "openai" | "custom"
	APIKey        string
	Model         string // provider-specific selector, e.g. "haiku", "sonnet-5", "luna", "terra"
	CustomBaseURL string
	CustomModel   string

	// Web search
	WebSearchEnabled  bool
	WebSearchProvider string
	WebSearchAPIKey   string // empty = keep stored
	WebSearchBaseURL  string
}

// UpsertAIAgent creates or updates the agent config for a product
// (ProductID nil = the default agent, used for General).
type UpsertAIAgent struct {
	ProductID        *int
	Description      string
	Instructions     string
	Enabled          bool
	WebSearchEnabled bool
}

// SaveIdeaBrief attaches Vora's brief to a post. Content must already carry
// the {{submitter_email}} token instead of a real address.
type SaveIdeaBrief struct {
	PostID          int
	Content         string
	Transcript      string // JSON []entity.AIMessage, already email-scrubbed
	SubmitterUserID int
}

// AIChatCompletion calls the tenant's configured provider with a system
// prompt and conversation, returning the assistant's reply.
type AIChatCompletion struct {
	System    string
	Messages  []entity.AIMessage
	MaxTokens int

	Result string
}

// AIWebSearch runs a web search through the tenant's configured provider
// (Serper.dev or a self-hosted SearXNG). Results are used only as untrusted
// grounding context for Vora — never executed or acted on.
type AIWebSearch struct {
	Query string

	Result []AIWebSearchResult
}

type AIWebSearchResult struct {
	Title   string
	URL     string
	Snippet string
}
